libs/vm/src/analyser.cpp

```diff
@@ -357,25 +390,438 @@ bool Analyser::Analyse(BlockNodePtr const &root, std::vector<std::string> &error
     file_errors_array_.clear();
     return false;
   }
-
   errors.clear();
   return true;
 }
 
+void Analyser::EnforceLedgerRestrictions(BlockNodePtr const &block_node)
+{
+  LedgerRestrictionMetadata metadata;
+
+  ValidateBlock(block_node, metadata);
+
+  CheckInitFunctionUnique(metadata);
+  if (CheckSynergeticFunctionsPresentAndUnique(metadata))
+  {
+    CheckSynergeticContract(metadata);
+  }
+}
+
+void Analyser::ValidateBlock(BlockNodePtr const &block_node, LedgerRestrictionMetadata &metadata)
+{
+  for (NodePtr const &child : block_node->block_children)
+  {
+    switch (child->node_kind)
+    {
+    case NodeKind::File:
+    {
+      BlockNodePtr file_node = ConvertToBlockNodePtr(child);
+      filename_              = file_node->text;
+      ValidateBlock(file_node, metadata);
+      break;
+    }
+    case NodeKind::ContractDefinition:
+    case NodeKind::StructDefinition:
+    {
+      ExpressionNodePtr name_node = ConvertToExpressionNodePtr(child->children[0]);
+      TypePtr           type      = name_node->type;
+      if (type)
+      {
+        ValidateBlock(ConvertToBlockNodePtr(child), metadata);
+      }
+      break;
+    }
+    case NodeKind::ContractFunction:
+    {
+      ValidateFunctionAnnotations(child);
+      ValidateFunctionPrototype(child, metadata);
+      break;
+    }
+    case NodeKind::FreeFunctionDefinition:
+    case NodeKind::MemberFunctionDefinition:
+    {
+      ValidateFunctionAnnotations(ConvertToBlockNodePtr(child));
```

Comment:
Need to check that the function name node ( child->children[1] ) has a valid function, and ignore the node if not

Reply:
done